### PR TITLE
Add option for allowing variables as elements

### PR DESCRIPTION
**Pull request by brandonpayton**
Hi @calebmer,

This PR adds an option to babel-plugin-transform-jsx to allow variables to be used as elements (component element names).

I nested the node transformers within the JSXElement visitor function so the JSXElementName transformer could have access to the `path` object and use `path.scope.hasBinding` to determine whether or not an element name references a variable. It seemed better to restructure this way than to pass an additional `path` argument to multiple transformers. Hopefully, you feel the same way, but I can update and go the latter route if you prefer.

How does this look? Are there any other changes you'd like under this PR?

### CODE REVIEW

File: babel-plugin-transform-jsx/README.md [hunk 1]
```diff
@@ -304,6 +305,14 @@ function MyComponent() {
 
 In the transformed object instead of having the string `foo.bar` for `elementName`, it would instead reference `foo.bar` the object property.
 
+However, if you want to use component element names as shown in the React example above, you can set the `useVariables` option to enable support:
+
+```json
+{
+  "plugins": [["transform-jsx", { "useVariables": true }]]
```
Comment (calebmer) [inline babel-plugin-transform-jsx/README.md]: Call this `useLocalVariables` so it's a little more explicit about what it does.

File: babel-plugin-transform-jsx/lib/index.js [hunk 2]
```diff
@@ -154,21 +58,139 @@ export default function ({ types: t }) {
...
+        t.memberExpression(
+          JSXMemberExpression(node.object),
+          JSXMemberExpression(node.property)
+        )
+      )
+    })
+
+    const JSXElementName = transformOnType({
+      JSXIdentifier: useVariables
+        ? node => path.scope.hasBinding(node.name) ? t.identifier(node.name) : JSXIdentifier(node)
```
Comment (calebmer) [inline babel-plugin-transform-jsx/lib/index.js]: My initial idea was to use a regular expression option and if it passed a variable reference would be used with the first capture grouped (if specified as the variable name) otherwise a string would be used.

For example `^[A-Z]` would make any React-styled component with the first letter as uppercase a variable, or `^&(.*)` which would make anything starting with an ampersand a variable (the following characters would be the name).

Could you make a quick argument for doing it this way? I think it may be a bit confusing. In addition sometimes when writing forms I have an `input` variable in scope which may be confusing. I think I'm cool with it, I just want a justification for it.

**Comment by calebmer** [comment]
Thanks for the PR! If you want to add your name to the contributors field, feel free!

I really like the closure refactor, that's something I should have just done in the first place 😊

I made a couple other comments, once they get resolved I'll merge in, do a small refactor of `initJSX` to work nicely with the closure, and publish.

File: babel-plugin-transform-jsx/README.md (hunk 1 above)
Comment (brandonpayton) [inline babel-plugin-transform-jsx/README.md]: The option name was bothering me as well. What do you think of `allowComponentElements` to be more specific about the what the feature actually is? In addition, it would be possible to use global variables if we opt for your approach of pattern-matching component element names.

File: babel-plugin-transform-jsx/lib/index.js (hunk 2 above)
Comment (brandonpayton) [inline babel-plugin-transform-jsx/lib/index.js]: TLDR: I now prefer your solution and propose option names and default at the end.

It was just the way I naturally thought of the feature.
1. Have component references
2. Want to reference them as elements
3. Need to differentiate between components and normal elements
4. Two ideas
  1. List normal elements and check for inclusion in the list -- would limit users to recognized list or require them to maintain their own list
  2. Look for variables in scope
5. Chose the latter as more maintainable

There was a little more to it, but that was my basic reasoning. After hearing your idea, I'm not sure which I prefer. I'll list some pros and cons for each.

# Check for bound names

* *Pros*
  * Requires no further configuration, just uses bound names if they're in scope
  * Based firmly in the reality of the scope (except global), rather than external naming rules
    * Is this ultimately more intuitive?
* *Cons*
  * Variables named after HTML elements will mask use of the element names
    * Would beg for a lint rule to warn about these cases
  * Denies the ability to reference global components
    * We could argue this is good, but I think it is bad to remove the choice.


# Use names that match a pattern

* *Pros*
  * Configuration makes an intentional statement about how components should be named
  * Can naturally avoid overlap with HTML element names by choosing a pattern that does not match HTML element names. Examples:
    * First letter uppercase
    * First character an ampersand (which I didn't know was permissable, but I'm trusting you :) )
  * Would not artificially deny use of global components
  * React appears to rely on naming convention in my tests. Lowercase names were not recognized as components, regardless of whether the lowercase names were valid HTML element names.
* *Cons*
  * Not based in the reality of the scope, can reference undeclared names
    * Would beg for a lint rule to warn about these cases

There are probably more, but this is enough for me.

# Thoughts

I prefer your solution because it allows clear separation of component and normal elements in the element namespace. In addition, I like the idea of uppercase indicating components and lowercase indicating normal elements and propose that as the default. What do you think?

If we go this route, I think we should have two options, one to enable the feature and the other to specify a custom naming pattern.

Following the proposal in my separate comment about option naming, what do you think of the following?
* allowComponentElements
* componentElementNamePattern

The latter feels a too wordy to me, but I like its specificity.

Comment (calebmer) [inline babel-plugin-transform-jsx/lib/index.js]: Great analysis, and you might actually be right about the ampersand thing 😊

My original idea was a little more ducktyp-ee. Have one configuration option named `variable` (to match the design of `module` and `function`, it could be `useVariable`). If the value were `true` we use the default regex (probably `^[A-Z]` as you suggest). If it's a string we use that as the regex. By default it would be `false`.

**Comment by brandonpayton** [comment]
Thanks for your comments. I've updated the code and README in response.

I just noticed that your last comment mentions the singular option names `variable` and `useVariable` where we were previously using a plural "variables". Let me know if this is intentional, and I'll change it.

How does the latest look?

**Comment by calebmer** [comment]
Looks great! Thanks so much for the pr 🙌 🎉

**Comment by calebmer** [comment]
This feature is now published at version `1.2.0`.

**Comment by brandonpayton** [comment]
Great! Thanks for merging. I'm excited to use this feature.

**Comment by brandonpayton** [comment]
Nice README edits too.
